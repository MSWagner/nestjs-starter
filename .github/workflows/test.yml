name: Test

on: [push, pull_request]

jobs:
  test:
    name: Test
    runs-on: ubuntu-latest

    steps:
      - uses: actions/checkout@v2

      - name: Build the docker-compose stack
        run: docker-compose -f 'docker-compose-test.yml' up -d
      - name: Check running containers
        run: docker ps -a
      - name: Check logs
        run: docker logs ananas-dark-backend-postgres-test

      - name: Use Node.js
        uses: actions/setup-node@v1
        with:
          node-version: '10.13.0'
      - name: Install dependencies
        run: yarn
      - name: Build App
        run: yarn build
      - name: Run database migrations
        run: yarn db:migrate:test
      - name: Start tests
        run: yarn test