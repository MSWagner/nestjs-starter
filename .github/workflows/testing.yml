name: Testing

on: [push, pull_request]

jobs:
  test-local:
    name: Testing Local
    runs-on: ubuntu-latest

    steps:
      - uses: actions/checkout@v2

      - name: Build the docker-compose stack
        run: docker-compose -f 'docker-compose.test.yml' up -d
      - name: Check running containers
        run: docker ps -a
      - name: Check logs
        run: docker logs project-name-postgres-test-standalone

      - name: Use Node.js
        uses: actions/setup-node@v1
        with:
          node-version: '18'
      - name: Install dependencies
        run: yarn
      - name: Build App
        run: yarn build
      - name: Run database migrations
        run: yarn db:migrate:test
      - name: Start tests
        run: yarn test

  test-docker:
    name: Testing Docker Container
    runs-on: ubuntu-latest

    steps:
      - uses: actions/checkout@v2

      - name: Use Node.js
        uses: actions/setup-node@v1
        with:
          node-version: '18'
      
      - name: Create docker network
        run: docker network create project-name-network-dev
      - name: Start Docker Dev Container
        run: yarn docker:up:hidden
      - name: Check running containers
        run: docker ps -a
      - name: Check logs
        run: docker logs project-name-postgres-test

      - name: Run database migrations
        run: yarn docker:db:migrate

      - name: Start tests in docker container
        run: yarn docker:test