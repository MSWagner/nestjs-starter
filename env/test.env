NODE_ENV=test

PGHOST=localhost
PGPORT=5431
PGUSER=root
PGPASSWORD=toor
PGDATABASE=project-name-test-db