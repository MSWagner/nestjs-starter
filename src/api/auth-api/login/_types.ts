import { IsString, MinLength } from "class-validator";
import { ApiProperty } from "@nestjs/swagger";

export class LoginDto {
    @IsString()
    @MinLength(3)
    @ApiProperty({ minLength: 3 })
    readonly username: string;

    @IsString()
    @MinLength(6)
    @ApiProperty({ minLength: 6 })
    readonly password: string;
}

export class LoginResponse {
    @ApiProperty()
    readonly tokenType: string;

    @ApiProperty()
    readonly expiresIn: number;

    @ApiProperty({ format: "uuid" })
    refreshToken: string;

    @ApiProperty({ format: "uuid" })
    accessToken: string;
}
