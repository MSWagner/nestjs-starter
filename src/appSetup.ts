import compression from "compression";
import helmet from "helmet";
import rateLimit from "express-rate-limit";

import { INestApplication, ValidationPipe, VersioningType } from "@nestjs/common";

export async function setupApp(app: INestApplication): Promise<void> {
    app.setGlobalPrefix("api");
    app.enableVersioning({
        type: VersioningType.URI
    });

    app.use(compression());
    app.use(helmet());
    app.enableCors();
    app.use(
        rateLimit({
            windowMs: 15 * 60 * 1000, // 15 minutes
            max: 100 // limit each IP to 100 requests per windowMs
        })
    );

    app.useGlobalPipes(new ValidationPipe({ whitelist: true, forbidNonWhitelisted: true }));
}
