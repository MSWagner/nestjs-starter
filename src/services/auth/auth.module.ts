import { Module } from "@nestjs/common";

import { AuthService } from "./auth.service";

import { LocalStrategy } from "./strategies/local.strategy";
import { BearerStrategy } from "./strategies/bearer.strategy";

import { UserModule } from "../user/user.module";

import { PassportModule } from "@nestjs/passport";

import { accessTokenProviders } from "./providers/accessTokenProviders";
import { refreshTokenProviders } from "./providers/refreshTokenProviders";

@Module({
    imports: [UserModule, PassportModule],
    providers: [AuthService, LocalStrategy, BearerStrategy, ...accessTokenProviders, ...refreshTokenProviders],
    exports: [AuthService]
})
export class AuthModule {}
