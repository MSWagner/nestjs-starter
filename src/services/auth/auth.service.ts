import * as bcrypt from "bcrypt";
import moment from "moment";
import * as _ from "lodash";

import { Injectable, Inject } from "@nestjs/common";
import { Repository } from "typeorm";

import { UserService } from "../user/user.service";

import { User } from "../../entities/User.entity";
import { AccessToken } from "../../entities/AccessToken.entity";
import { RefreshToken } from "../../entities/RefreshToken.entity";

import { CONFIG } from "../../config";

export interface IToken {
    accessToken: AccessToken;
    refreshToken: RefreshToken;
}

@Injectable()
export class AuthService {
    constructor(
        @Inject(CONFIG.database.defaultAccessTokenRepoName)
        private readonly accessTokenRepository: Repository<AccessToken>,

        @Inject(CONFIG.database.defaultRefreshTokenRepoName)
        private readonly refreshTokenRepository: Repository<RefreshToken>,

        private readonly userService: UserService
    ) {}

    async generateToken(user: User): Promise<{ accessToken: AccessToken; refreshToken: RefreshToken }> {
        const newRefreshToken = new RefreshToken();
        newRefreshToken.validUntil =
            CONFIG.auth.refreshTokenValidityMS > 0
                ? moment().add(CONFIG.auth.refreshTokenValidityMS, "milliseconds").toDate()
                : null;
        newRefreshToken.user = user;

        const accessToken = await this.generateAuthToken(user);
        const refreshToken = await this.refreshTokenRepository.save(newRefreshToken);

        return {
            accessToken,
            refreshToken
        };
    }

    async generateAuthToken(user: User): Promise<AccessToken> {
        const newAccessToken = new AccessToken();
        newAccessToken.validUntil = moment().add(CONFIG.auth.tokenValidity, "milliseconds").toDate();
        newAccessToken.user = user;

        const accessToken = await this.accessTokenRepository.save(newAccessToken);

        return accessToken;
    }

    async validateUser(username: string, password: string): Promise<User> {
        const user = await this.userService.findOne(username);

        if (_.isNil(user)) {
            return null;
        }

        if (!user.isActive) {
            return null;
        }

        const isValid = await bcrypt.compare(password, user.passwordHash);

        return isValid ? user : null;
    }

    async validateToken(token: string): Promise<User> {
        const accessToken = await this.accessTokenRepository.findOne(token, { relations: ["user"] });

        if (_.isNil(accessToken)) {
            return null;
        }

        const isValid = moment().isSameOrBefore(accessToken.validUntil);

        if (!isValid) {
            await this.accessTokenRepository.delete(accessToken);
        }

        if (!accessToken.user.isActive) {
            return null;
        }

        return isValid ? accessToken.user : null;
    }

    async register(username: string, password: string): Promise<User> {
        return this.userService.createUser(username, password);
    }

    async login(user: User): Promise<IToken> {
        return this.generateToken(user);
    }

    async refreshAuthToken(token: string): Promise<IToken> {
        const refreshToken = await this.refreshTokenRepository.findOne(token, { relations: ["user"] });

        const isValid =
            !_.isNil(refreshToken) &&
            !_.isNil(refreshToken.user) &&
            (_.isNil(refreshToken.validUntil) || moment().isSameOrBefore(refreshToken.validUntil));

        if (!isValid && !_.isNil(refreshToken)) {
            await this.refreshTokenRepository.delete(refreshToken);
        }

        return !isValid
            ? null
            : {
                  accessToken: await this.generateAuthToken(refreshToken.user),
                  refreshToken
              };
    }
}
