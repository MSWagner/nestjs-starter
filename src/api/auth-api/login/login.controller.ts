import { Controller, Request, Post, UseGuards } from "@nestjs/common";
import { AuthGuard } from "@nestjs/passport";
import { ApiBody, ApiCreatedResponse, ApiTags } from "@nestjs/swagger";

import { AuthService } from "../../../services/auth/auth.service";
import { LoginDto, LoginResponse } from "./_types";

import { CONFIG } from "../../../config";

import { Permissions } from "../../../services/auth/permissions/permission.decorator";
import { PermissionsGuard } from "../../../services/auth/permissions/permission.guard";
import { PermissionScope } from "../../../entities/Permission.entity";

@ApiTags("auth")
@UseGuards(AuthGuard("local"))
@Controller({ path: "auth/login", version: "1" })
export class LoginController {
    constructor(private readonly authService: AuthService) {}

    @Post()
    @UseGuards(PermissionsGuard)
    @Permissions(PermissionScope.User)
    @ApiBody({ type: LoginDto })
    @ApiCreatedResponse({ description: "The credentials has been successfully created.", type: LoginResponse })
    async handler(@Request() req): Promise<LoginResponse> {
        const tokens = await this.authService.login(req.user);

        return {
            tokenType: "Bearer",
            expiresIn: CONFIG.auth.tokenValidity,
            accessToken: tokens.accessToken.token,
            refreshToken: tokens.refreshToken.token
        };
    }
}
