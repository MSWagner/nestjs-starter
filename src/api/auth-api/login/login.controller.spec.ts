import request from "supertest";
import { Test, TestingModule } from "@nestjs/testing";
import { INestApplication } from "@nestjs/common";

import { setupApp } from "../../../appSetup";

import { TestModule } from "../../../services/test/test.module";
import { TestService } from "../../../services/test/test.service";
import { LoginController } from "./login.controller";
import { AuthModule } from "../../../services/auth/auth.module";

import { AccessToken } from "../../../entities/AccessToken.entity";
import { RefreshToken } from "../../../entities/RefreshToken.entity";

import * as fixtures from "../../../services/test/fixtures";

describe("Login Controller", () => {
    let app: INestApplication;
    let testService: TestService;

    beforeAll(async () => {
        const module: TestingModule = await Test.createTestingModule({
            controllers: [LoginController],
            imports: [TestModule, AuthModule]
        }).compile();

        testService = module.get<TestService>(TestService);

        app = module.createNestApplication();
        await setupApp(app);
        await app.init();
    });

    beforeEach(async () => {
        await testService.reloadFixtures();
    });

    afterAll((done) => {
        testService.connectionManager.connection.close();
        done();
    });

    it("/POST auth/login - should login user1", async () => {
        const body = {
            username: fixtures.user1.username,
            password: "testPassword"
        };

        const refreshTokenCountBefore = await RefreshToken.count({ user: fixtures.user1 });
        expect(refreshTokenCountBefore).toEqual(1);

        const accessTokenCountBefore = await AccessToken.count({ user: fixtures.user1 });
        expect(accessTokenCountBefore).toEqual(1);

        const response = await request(app.getHttpServer()).post("/api/v1/auth/login").send(body).expect(201);

        const refreshTokenCountAfter = await RefreshToken.count({ user: fixtures.user1 });
        expect(refreshTokenCountAfter).toEqual(refreshTokenCountBefore + 1);

        const accessTokenCountAfter = await AccessToken.count({ user: fixtures.user1 });
        expect(accessTokenCountAfter).toEqual(accessTokenCountBefore + 1);

        const credentials = testService.replaceValues(response.body, ["UUID", "DATE"]);
        expect(credentials).toMatchSnapshot("LoginUser1");
    });

    it("/POST auth/login - should throw error for missing username", async () => {
        const body = {
            password: "testPassword"
        };

        const refreshTokenCountBefore = await RefreshToken.count({ user: fixtures.user1 });
        expect(refreshTokenCountBefore).toEqual(1);

        const accessTokenCountBefore = await AccessToken.count({ user: fixtures.user1 });
        expect(accessTokenCountBefore).toEqual(1);

        const response = await request(app.getHttpServer()).post("/api/v1/auth/login").send(body).expect(401);

        const refreshTokenCountAfter = await RefreshToken.count({ user: fixtures.user1 });
        expect(refreshTokenCountAfter).toEqual(refreshTokenCountBefore);

        const accessTokenCountAfter = await AccessToken.count({ user: fixtures.user1 });
        expect(accessTokenCountAfter).toEqual(accessTokenCountBefore);

        const credentials = testService.replaceValues(response.body, ["UUID", "DATE"]);
        expect(credentials).toMatchSnapshot("LoginMissingUsernameError");
    });

    it("/POST auth/login - should throw error for missing password", async () => {
        const body = {
            username: fixtures.user1.username
        };

        const refreshTokenCountBefore = await RefreshToken.count({ user: fixtures.user1 });
        expect(refreshTokenCountBefore).toEqual(1);

        const accessTokenCountBefore = await AccessToken.count({ user: fixtures.user1 });
        expect(accessTokenCountBefore).toEqual(1);

        const response = await request(app.getHttpServer()).post("/api/v1/auth/login").send(body).expect(401);

        const refreshTokenCountAfter = await RefreshToken.count({ user: fixtures.user1 });
        expect(refreshTokenCountAfter).toEqual(refreshTokenCountBefore);

        const accessTokenCountAfter = await AccessToken.count({ user: fixtures.user1 });
        expect(accessTokenCountAfter).toEqual(accessTokenCountBefore);

        const credentials = testService.replaceValues(response.body, ["UUID", "DATE"]);
        expect(credentials).toMatchSnapshot("LoginMissingPasswordError");
    });

    it("/POST auth/login - should return an error for a not existing username", async () => {
        const body = {
            username: "notExisting",
            password: "testPassword"
        };

        const refreshTokenCountBefore = await RefreshToken.count({ user: fixtures.user1 });
        expect(refreshTokenCountBefore).toEqual(1);

        const accessTokenCountBefore = await AccessToken.count({ user: fixtures.user1 });
        expect(accessTokenCountBefore).toEqual(1);

        const response = await request(app.getHttpServer()).post("/api/v1/auth/login").send(body).expect(401);

        const refreshTokenCountAfter = await RefreshToken.count({ user: fixtures.user1 });
        expect(refreshTokenCountAfter).toEqual(refreshTokenCountBefore);

        const accessTokenCountAfter = await AccessToken.count({ user: fixtures.user1 });
        expect(accessTokenCountAfter).toEqual(accessTokenCountBefore);

        const credentials = testService.replaceValues(response.body, ["UUID", "DATE"]);
        expect(credentials).toMatchSnapshot("LoginNotExistingUserError");
    });

    it("/POST auth/login - should return an error for a wrong password", async () => {
        const body = {
            username: fixtures.user1.username,
            password: "wrongPassword"
        };

        const refreshTokenCountBefore = await RefreshToken.count({ user: fixtures.user1 });
        expect(refreshTokenCountBefore).toEqual(1);

        const accessTokenCountBefore = await AccessToken.count({ user: fixtures.user1 });
        expect(accessTokenCountBefore).toEqual(1);

        const response = await request(app.getHttpServer()).post("/api/v1/auth/login").send(body).expect(401);

        const refreshTokenCountAfter = await RefreshToken.count({ user: fixtures.user1 });
        expect(refreshTokenCountAfter).toEqual(refreshTokenCountBefore);

        const accessTokenCountAfter = await AccessToken.count({ user: fixtures.user1 });
        expect(accessTokenCountAfter).toEqual(accessTokenCountBefore);

        const credentials = testService.replaceValues(response.body, ["UUID", "DATE"]);
        expect(credentials).toMatchSnapshot("LoginWrongPasswordError");
    });

    it("/POST auth/login - should return an error for a missing permission scope (user2)", async () => {
        const body = {
            username: fixtures.user2.username,
            password: "testPassword"
        };

        const refreshTokenCountBefore = await RefreshToken.count({ user: fixtures.user1 });
        expect(refreshTokenCountBefore).toEqual(1);

        const accessTokenCountBefore = await AccessToken.count({ user: fixtures.user1 });
        expect(accessTokenCountBefore).toEqual(1);

        const response = await request(app.getHttpServer()).post("/api/v1/auth/login").send(body).expect(403);

        const refreshTokenCountAfter = await RefreshToken.count({ user: fixtures.user1 });
        expect(refreshTokenCountAfter).toEqual(refreshTokenCountBefore);

        const accessTokenCountAfter = await AccessToken.count({ user: fixtures.user1 });
        expect(accessTokenCountAfter).toEqual(accessTokenCountBefore);

        const credentials = testService.replaceValues(response.body, ["UUID", "DATE"]);
        expect(credentials).toMatchSnapshot("MissingPermissionScopeUser2");
    });
});
