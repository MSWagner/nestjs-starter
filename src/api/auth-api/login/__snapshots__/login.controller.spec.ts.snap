// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Login Controller /POST auth/login - should login user1: LoginUser1 1`] = `
{
  "accessToken": "accessToken",
  "expiresIn": 86400000,
  "refreshToken": "refreshToken",
  "tokenType": "Bearer",
}
`;

exports[`Login Controller /POST auth/login - should return an error for a missing permission scope (user2): MissingPermissionScopeUser2 1`] = `
{
  "error": "Forbidden",
  "message": "Forbidden resource",
  "statusCode": 403,
}
`;

exports[`Login Controller /POST auth/login - should return an error for a not existing username: LoginNotExistingUserError 1`] = `
{
  "message": "Unauthorized",
  "statusCode": 401,
}
`;

exports[`Login Controller /POST auth/login - should return an error for a wrong password: LoginWrongPasswordError 1`] = `
{
  "message": "Unauthorized",
  "statusCode": 401,
}
`;

exports[`Login Controller /POST auth/login - should throw error for missing password: LoginMissingPasswordError 1`] = `
{
  "message": "Unauthorized",
  "statusCode": 401,
}
`;

exports[`Login Controller /POST auth/login - should throw error for missing username: LoginMissingUsernameError 1`] = `
{
  "message": "Unauthorized",
  "statusCode": 401,
}
`;
